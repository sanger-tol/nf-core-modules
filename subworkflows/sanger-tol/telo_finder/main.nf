//
// MODULE IMPORT BLOCK
//
include { TELOMERE_REGIONS          } from '../../../modules/sanger-tol/telomere/regions/main'
include { GAWK                      } from '../../../modules/nf-core/gawk/main'
include { TELOMERE_WINDOWS          } from '../../../modules/sanger-tol/telomere/windows/main'
include { TELOMERE_EXTRACT          } from '../../../modules/sanger-tol/telomere/extract/main'

workflow TELO_FINDER {

    take:
    ch_reference        // Channel [ val(meta), path(fasta) ]
    ch_telomereseq      // Channel.of( telomere sequence )
    val_split_telomere  // bool

    main:
    ch_versions         = Channel.empty()


    //
    // MODULE: FINDS THE TELOMERIC SEQEUNCE IN REFERENCE
    //
    TELOMERE_REGIONS (
        ch_reference,
        ch_telomereseq
    )
    ch_versions         = ch_versions.mix(TELOMERE_REGIONS.out.versions)

    TELOMERE_REGIONS.out.telomere
        .map{ meta, file ->
            def new_meta = meta + [direction: 0]
            [new_meta, file]
        }
        .set { ch_full_telomere }

    //
    // MODULE: SPLIT THE TELOMERE FILE INTO 5' and 3' FILES
    //
    if (val_split_telomere) {

        GAWK (
            ch_full_telomere,
            [],
            true
        )
        ch_versions     = ch_versions.mix(GAWK.out.versions)

        //
        // LOGIC: COLLECT FILES AND ITERATE THROUGH
        //          CHANGE meta.id BASED ON THE FILE NAME
        //          THIS PRODUCES A TRIO OF CHANNELS: [meta], file
        //          THESE CAN BE 0 SIZE AND CAUSE ISSUES DOWNSTREAM
        //
        GAWK.out.output
            .flatMap { meta, files ->
                files
                    .findAll { file -> file.size() > 0 }
                    .collect { file ->
                        if (file.name.contains("direction.0")) {
                            new_meta = meta + [direction: 5]
                        }
                        if (file.name.contains("direction.1")) {
                            new_meta = meta + [direction: 3]
                        }
                        [new_meta, file]
                    }
            }
            .mix(ch_full_telomere)
            .set { ch_regions_for_extraction }


    } else {
        ch_regions_for_extraction  = ch_full_telomere
    }


    //
    // MODULE: GENERATES A WINDOWS FILE FROM THE ABOVE
    //
    TELOMERE_WINDOWS (
        ch_regions_for_extraction
    )
    ch_versions         = ch_versions.mix(TELOMERE_WINDOWS.out.versions)


    //
    // LOGIC: OUTPUT CAN HAVE SIZE 0 WHICH BREAKS gawk IN EXTRACT
    //          FILTER OUT THE 0 SIZE FILES
    //
    TELOMERE_WINDOWS.out.windows
        .filter { _meta, file ->
            file.size() > 0
        }
        .set { ch_filtered_windows_for_extraction  }

    //
    // MODULE: Extract the telomere data from the FIND_TELOMERE
    //          file and reformat into bed
    //
    TELOMERE_EXTRACT(
        ch_filtered_windows_for_extraction
    )
    ch_versions         = ch_versions.mix(TELOMERE_EXTRACT.out.versions)


    TELOMERE_EXTRACT.out.bedgraph
        .map { meta, bedgraph ->
            [ meta - meta.subMap("direction"), bedgraph ]
        }
        .groupTuple(by: 0, sort: true)
        .set { ch_telo_bedgraphs }


    emit:
    bedgraph_file       = ch_telo_bedgraphs    // Used in pretext_graph
    versions            = ch_versions
}
