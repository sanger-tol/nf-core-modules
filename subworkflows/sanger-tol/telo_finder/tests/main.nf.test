nextflow_workflow {

    name "Test Subworkflow TELO_FINDER"
    script "../main.nf"
    config "./nextflow.config"
    workflow "TELO_FINDER"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/telo_finder"
    tag "telomere/regions"
    tag "telomere/windows"
    tag "telomere/extract"
    tag "subworkflows/../../modules/nf-core/gawk"
	tag "subworkflows/../../modules/nf-core/gunzip"



    test("idFanCani4 - no split - fasta") {
        when {
            params {
                gawk_args       = "'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } {print > \"direction.\"\$3\".split.telomere\"}'"
                gawk_suffix     = "split.telomere"
            }

            workflow {
                """
                input[0] = [ [id: "idFanCani4"], file("/lustre/scratch124/tol/teams/tola/users/dp24/nf-core-modules/idFanCani4.hap1.1.primary.curated.part_002.fa")]
                input[1] = "TTAGG"
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("idFanCani4 - split - fasta") {
        when {
            params {
                gawk_args       = "'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } {print > \"direction.\"\$3\".split.telomere\"}'"
                gawk_suffix     = "split.telomere"
            }

            workflow {
                """
                input[0] = [ [id: "idFanCani4"], file("/lustre/scratch124/tol/teams/tola/users/dp24/nf-core-modules/idFanCani4.hap1.1.primary.curated.part_002.fa")]
                input[1] = "TTAGG"
                input[2] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("idFanCani4 - no split - fasta - stub") {
        when {
            params {
                gawk_args       = "'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } {print > \"direction.\"\$3\".split.telomere\"}'"
                gawk_suffix     = "split.telomere"
            }

            workflow {
                """
                input[0] = [ [id: "idFanCani4"], file("/lustre/scratch124/tol/teams/tola/users/dp24/nf-core-modules/idFanCani4.hap1.1.primary.curated.part_002.fa")]
                input[1] = "TTAGG"
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }
}
