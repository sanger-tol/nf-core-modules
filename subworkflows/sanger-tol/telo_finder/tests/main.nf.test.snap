{
    "idFanCani4 - no split - fasta - stub": {
        "content": [
            {
                "0": [
                    [
                        "idFanCani4_telomere.bedgraph:md5,f8df4805d7f8948a5bc74c674965d064"
                    ]
                ],
                "1": [
                    "versions.yml:md5,02bd065343bd2db9fbe8c9eb341f1fb7",
                    "versions.yml:md5,29f570909fd629fc4a80db0039b84506",
                    "versions.yml:md5,2db8465f80039d90274c185879ae48bc"
                ],
                "bedgraph_file": [
                    [
                        "idFanCani4_telomere.bedgraph:md5,f8df4805d7f8948a5bc74c674965d064"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,02bd065343bd2db9fbe8c9eb341f1fb7",
                    "versions.yml:md5,29f570909fd629fc4a80db0039b84506",
                    "versions.yml:md5,2db8465f80039d90274c185879ae48bc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-10-21T16:07:21.368077078"
    },
    "idFanCani4 - no split - fasta": {
        "content": [
            {
                "0": [
                    [
                        "idFanCani4_telomere.bedgraph:md5,f8df4805d7f8948a5bc74c674965d064"
                    ]
                ],
                "1": [
                    "versions.yml:md5,02bd065343bd2db9fbe8c9eb341f1fb7",
                    "versions.yml:md5,29f570909fd629fc4a80db0039b84506",
                    "versions.yml:md5,2db8465f80039d90274c185879ae48bc"
                ],
                "bedgraph_file": [
                    [
                        "idFanCani4_telomere.bedgraph:md5,f8df4805d7f8948a5bc74c674965d064"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,02bd065343bd2db9fbe8c9eb341f1fb7",
                    "versions.yml:md5,29f570909fd629fc4a80db0039b84506",
                    "versions.yml:md5,2db8465f80039d90274c185879ae48bc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-10-21T16:06:02.528534255"
    },
    "idFanCani4 - split - fasta": {
        "content": [
            [
                [
                    "idFanCani4_telomere.bedgraph:md5,f8df4805d7f8948a5bc74c674965d064",
                    "idFanCani4_3P_telomere.bedgraph:md5,ae30aa41f6e9c53df160ddb17ecccd09",
                    "idFanCani4_5P_telomere.bedgraph:md5,ac00ddc89af099cfdd820495cd8c8c76"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-10-21T16:06:44.589353231"
    }
}